src/v/storage/segment_reader.cc

```diff
@@ -244,4 +244,87 @@ void segment_reader_handle::operator=(segment_reader_handle&& rhs) noexcept {
     _hook.swap_nodes(rhs._hook);
 }
 
+concat_segment_data_source_impl::concat_segment_data_source_impl(
+  std::vector<ss::lw_shared_ptr<segment>> segments,
+  size_t start_pos,
+  size_t end_pos,
+  ss::io_priority_class priority_class)
+  : _segments{std::move(segments)}
+  , _current_pos{_segments.begin()}
+  , _start_pos{start_pos}
+  , _end_pos{end_pos}
+  , _priority_class{priority_class} {}
+
+ss::future<ss::temporary_buffer<char>> concat_segment_data_source_impl::get() {
+    if (!_current_stream) {
+        _current_stream = co_await next_stream();
+    }
+
+    ss::temporary_buffer<char> buf = co_await _current_stream->read();
+    while (buf.empty() && _current_pos != _segments.end()) {
+        _current_stream = co_await next_stream();
+        buf = co_await _current_stream->read();
+    }
+
+    co_return buf;
+}
+
+ss::future<ss::input_stream<char>>
+concat_segment_data_source_impl::next_stream() {
+    if (_current_handle) {
+        co_await _current_handle->close();
+    }
+
+    vlog(
+      stlog.trace,
+      "opening stream for segment index {}, total segments: {}",
+      std::distance(_segments.begin(), _current_pos),
+      _segments.size());
+    auto segment = *_current_pos;
+    size_t start = 0;
+    size_t end = segment->file_size();
+
+    if (_current_pos == _segments.begin()) {
+        start = _start_pos;
+    }
+
+    if (std::distance(_current_pos, _segments.end()) == 1) {
```

Comment:
is this test effectively `if current_pos == std::prev(_segments.end())`?

Reply:
Yes, this is the check, we are testing that `_current_pos` points to the penultimate position. Is the `std::prev` style preferred? If so I can change it.

Reply:
I changed it to use `std::prev` since your version seems easier to understand . Are there are other advantages to using `std::prev` over distance?

Reply:
>I changed it to use std::prev since your version seems easier to understand . Are there are other advantages to using std::prev over distance?

mm, i don't really know. probably not for containers where std::distance is O(1)